src/components/user.js

```diff
@@ -15,69 +25,136 @@ export default class User extends React.Component {
         const { info = {}, address = {}, company = {} } = this.props;
 
         this.state = {
-            currentUser: {
+            currentUser:         {
                 info,
                 address,
                 company
             },
-            showAddress: false,
-            showCompany: false,
-            wantEdit:    false
+            unmodifiedUser:      {
+                info,
+                address,
+                company
+            },
+            formErrors:          {
+                name:  '',
+                email: '',
+                phone: ''
+            },
+            showAddress:         false,
+            showCompany:         false,
+            wantEdit:            false,
+            forceErrorDuplicate: false
         };
+    }
 
-        this.handleClickAddress = this.handleClickAddress.bind(this);
-        this.handleClickCompany = this.handleClickCompany.bind(this);
-        this.handleChange       = this.handleChange.bind(this);
-        this.handleClickEdit    = this.handleClickEdit.bind(this);
-        this.handleClickSubmit  = this.handleClickSubmit.bind(this);
+    _validateField (fieldName, value) {
+        switch (fieldName) {
+            case FIELD_NAMES.name:
+                return value ? '' : ERRORS.nameInvalid;
+            case FIELD_NAMES.phone:
+                return value && MATCH_PHONE.test(value) ? '' : ERRORS.phoneInvalid;
+            case FIELD_NAMES.email:
+                return value && MATCH_EMAIL.test(value) ? '' : ERRORS.emailInvalid;
+            default:
+                return '';
+        }
     }
 
-    handleClickAddress (e) {
+    _validateFields (user) {
+        return {
+            [FIELD_NAMES.name]:  this._validateField(FIELD_NAMES.name, user.name),
+            [FIELD_NAMES.phone]: this._validateField(FIELD_NAMES.phone, user.phone),
+            [FIELD_NAMES.email]: this._validateField(FIELD_NAMES.email, user.email),
+        };
+    }
+
+    _isUserInfoValid (validInfo) {
+        return !validInfo[FIELD_NAMES.name] &&
+               !validInfo[FIELD_NAMES.phone] &&
+               !validInfo[FIELD_NAMES.email];
+    }
+
+    _handleClickAddress = e => {
         this.setState({ showAddress: !this.state.showAddress });
         e.preventDefault();
-    }
+    };
 
-    handleClickCompany (e) {
+    _handleClickCompany = e => {
         this.setState({ showCompany: !this.state.showCompany });
         e.preventDefault();
-    }
+    };
+
+    _handleClickEdit = e => {
```

Comment:
Let's create two separate methods:
- for start editing 
- for clicking on `Save` button